### PR TITLE
Rework Rust runtime and builder

### PR DESCRIPTION
**Pull request by ecklf**
This PR reworks the builder and runtime to support the latest AWS lambda runtime which changed [almost everything](https://github.com/awslabs/aws-lambda-rust-runtime/releases/tag/v0.3.0#:~:text=New%3A-,Almost%20everything,-!). Examples can be found in the `examples` folder.

Optimizations that will be considered in follow-up PRs:

* Improve build caching by looking at https://github.com/Swatinem/rust-cache.
* Merge handlers into one lambda. [Build scripts](https://doc.rust-lang.org/cargo/reference/build-scripts.html#build-scripts) might be a good solution for that.
* CI setup

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -16,55 +16,59 @@ First, you'll need a `vercel.json` file in your project:
...
+      "runtime": "vercel-rust@4.0.0"
     }
   }
 }
 ```
 
-A Vercel Function will be created for every file that matches `api/**/*.rs`. Next, you can create a new Function `api/user.rs`:
+A Vercel Function will be created for every file that matches `api/**/*.rs`.
+
+<!-- todo annotation for filename? -->
```
Comment (dglsparsons) [inline README.md]: 🤔

File: README.md [hunk 2]
```diff
@@ -16,55 +16,57 @@ First, you'll need a `vercel.json` file in your project:
 {
   "functions": {
     "api/**/*.rs": {
-      "runtime": "vercel-rust@3.1.2"
+      "runtime": "vercel-rust@4.0.0"
```
Comment (ecklf) [inline README.md]: This is not yet published. I have a tmp package set up for the latest builder: [ecklf-tmp-runtime-test@1.0.83](https://www.npmjs.com/package/ecklf-tmp-runtime-test)

File: README.md [hunk 3]
```diff
@@ -16,55 +16,57 @@ First, you'll need a `vercel.json` file in your project:
...
+#[tokio::main]
+async fn main() -> Result<(), LambdaError> {
+    run(handler).await?;
+    Ok(())
 }
 
-// Start the runtime with the handler
-fn main() -> Result<(), Box<dyn Error>> {
-	Ok(lambda!(handler))
+pub async fn handler(_req: ProxyRequest) -> Result<impl IntoResponse, ProxyError> {
```
Comment (dglsparsons) [inline README.md]: See our chat about `Proxy` being confusing naming. Think we can remove it and just have `Request / Response / Error`

File: README.md [hunk 4]
```diff
@@ -16,55 +16,57 @@ First, you'll need a `vercel.json` file in your project:
...
-name = "util"
-path = "_util.rs"
+# You can specify a library for shared logic here (optional)
+# [lib]
+# path = "src-rs/lib.rs"
+
+# Each handler has to be specified as [[bin]]
+[[bin]]
+name = "handler"
+path = "api/handler.rs"
```
Comment (dglsparsons) [inline README.md:75]: I think we need to find a solution to doing this via `build.rs` or similar. Or maybe some regular `src/main.rs` that you drop a macro in, and it pulls all the handlers from `/api/...`

File: README.md [hunk 5]
```diff
@@ -73,15 +75,15 @@ Furthermore, more system dependencies can be installed at build time with the pr
...
+During local development with `vercel dev`, ensure `rust` and `cargo` are already installed and available in your `PATH`, since they will not be installed automatically. The recommended way to install is with [rustup](https://rustup.rs/).
 
 ## Contributing
 
 Since this project contains both Rust and Node.js code, you need to install the relevant dependencies. If you're only working on the JavaScript side, you only need to install those dependencies (and vice-versa).
 
-```sh
+```
 # install node dependencies
 npm install
```
Comment (dglsparsons) [inline README.md:99]: Is the rest of the README still correct? i.e. about Cargo workspaces being unsupported?

File: demo/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,28 @@
+[package]
+name = "runtime_demo"
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
Comment (dglsparsons) [inline demo/Cargo.toml]: we should remove this comment i think.

File: demo/vercel.json [hunk 7]
```diff
@@ -0,0 +1,7 @@
+{
+  "functions": {
+    "api/**/*.rs": {
+      "runtime": "ecklf-tmp-runtime-test@1.0.83"
```
Comment (dglsparsons) [inline demo/vercel.json]: make sure this is changed later.

File: package.json [hunk 8]
```diff
@@ -10,36 +10,22 @@
...
-  },
-  "prettier": {
-    "arrowParens": "always",
-    "singleQuote": true,
-    "tabWidth": 2,
-    "trailingComma": "none"
+    "@types/node": "^18.11.18",
+    "@vercel/build-utils": "^6.2.1",
+    "@vercel/routing-utils": "^2.1.8",
+    "rimraf": "^4.1.1"
```
Comment (dglsparsons) [inline package.json]: we should probably keep prettier?

File: src/lib/runtime.ts [hunk 9]
```diff
@@ -0,0 +1,19 @@
+import {
+  BuildOptions,
+  BuildResultV3,
+  Files,
+  PrepareCacheOptions,
+  ShouldServeOptions,
+  StartDevServerOptions,
+  StartDevServerResult,
+} from "@vercel/build-utils/dist";
```
Comment (dglsparsons) [inline src/lib/runtime.ts]: Are these all types? Can we use type imports?

File: tsconfig.json [hunk 10]
```diff
@@ -1,20 +1,106 @@
...
-    "noUnusedLocals": true,
-    "noUnusedParameters": true,
-    "outDir": "dist",
-    "sourceMap": false,
-    "strict": true,
-    "target": "esnext",
-    "types": ["node"],
-    "skipLibCheck": true
-  }
+    /* Visit https://aka.ms/tsconfig to read more about this file */
```
Comment (dglsparsons) [inline tsconfig.json]: can we remove comments?

File: vercel_runtime/Cargo.toml [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+tracing = { version = "0.1", features = ["log"] }
+tracing-subscriber = { version = "0.3", default-features = false, features = ["fmt"] }
+serde = { version = "1.0.145", features = ["derive"] }
+serde_json = { version = "1.0.86", features = ["raw_value"] }
+serde_derive = "1.0.9"
+tower-http = { version = "0.3.4", features = ["cors"] }
+tower-service = "0.3.2"
+base64 = "0.10"
+bytes = "1.3.0"
+async-trait = "0.1.63"
```
Comment (dglsparsons) [inline vercel_runtime/Cargo.toml]: new line at eof

File: vercel_runtime/src/request.rs [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+use serde::de::{Deserializer, Error as DeError, MapAccess, Visitor};
+use serde_derive::Deserialize;
+use serde_json::Value;
+use std::{borrow::Cow, fmt, mem};
+
+/// Representation of a Vercel Lambda proxy event data
+#[doc(hidden)]
+#[derive(Deserialize, Debug, Default)]
+#[serde(rename_all = "camelCase")]
+pub(crate) struct VercelRequest<'a> {
```
Comment (dglsparsons) [inline vercel_runtime/src/request.rs:13]: wait, so ProxyRequest is a request, and VercelRequest is a request according to the proxy? is that backwards? 🤔

File: README.md (hunk 5 above)
Comment (ecklf) [reply README.md:99]: Yes and no, with this new approach you can use workspaces, however you'd need to setup redirects as we have file system based routing (I couldn't figure out how to modify the routes in the build script). Still need to update the README fully on that regard.

File: demo/vercel.json (hunk 7 above)
Comment (ecklf) [reply demo/vercel.json]: I'll be switching this out for a canary publish in a bit

File: README.md (hunk 5 above)
Comment (dglsparsons) [reply README.md:99]: Interesting. We should figure out how to make that work nicely.

File: vercel_runtime/src/request.rs (hunk 12 above)
Comment (ecklf) [reply vercel_runtime/src/request.rs:13]: `VercelEvent` is our entrypoint when a lambda is invoked. `VercelRequest` is the `body` of that

File: README.md (hunk 4 above)
Comment (ecklf) [reply README.md:75]: I thought of that as well, issue being that then rust analyzer doesn't pick up the files during development.

File: examples/nextjs/styles/globals.css [hunk 13]
```diff
@@ -0,0 +1,31 @@
+@tailwind base;
+
+:root {
+  @apply scroll-smooth font-display antialiased;
+}
+
+html {
+  @apply bg-gray-900;
```
Comment (dglsparsons) [inline examples/nextjs/styles/globals.css:8]: wtf?

**Review by dglsparsons — APPROVED** [review]

**Comment by abcd-ca** [comment]
Just did a quick smoke test. Much better, good job, thanks!

**Pull request merged**